src/Controllers/UserController.js

```diff
@@ -0,0 +1,222 @@
+  checkResetTokenValidity(username, token) {
+    return new Promise((resolve, reject) => {
+      return this.config.database.collection('_User').then(coll => {
+        return coll.findOne({
+          username: username,
+          _perishable_token: token,
+        }, (err, doc) => {
+          if (err || !doc) {
+            reject(err);
+          } else {
+            resolve(doc);
+          }
+        });
+      });
+    });
+  }
+  
+  getUserIfNeeded(user) {
+    if (user.username && user.email) {
+      return Promise.resolve(user);
+    }
+    var where = {};
+    if (user.username) {
+      where.username = user.username;
+    }
+    if (user.email) {
+      where.email = user.email;
+    }
+    
+    var query = new RestQuery(this.config, Auth.master(this.config), '_User', where);
+    return query.execute().then(function(result){
+      if (result.results.length != 1) {
+        return Promise.reject();
+      }
+      return result.results[0];
+    })
+  }
+  
+
+  sendVerificationEmail(user) {
+    if (!this.shouldVerifyEmails) {
+      return;
+    }
+    // We may need to fetch the user in case of update email
+    this.getUserIfNeeded(user).then((user) => {
+      const token = encodeURIComponent(user._email_verify_token);
+      const username = encodeURIComponent(user.username);
+      let link = `${this.config.verifyEmailURL}?token=${token}&username=${username}`;
+      let options = {
+        appName: this.config.appName,
+        link: link,
+        user: inflate('_User', user),
+      };
+      if (this.adapter.sendVerificationEmail) {
+        this.adapter.sendVerificationEmail(options);
+      } else {
+        this.adapter.sendMail(this.defaultVerificationEmail(options));
+      }
+    });
+  }
+  
+  setPasswordResetToken(email) {
+    var database = this.config.database;
+    var token = randomString(25);
+    return new Promise((resolve, reject) => {
+      return database.collection('_User').then(coll => {
+        // Need direct database access because verification token is not a parse field
+        return coll.findAndModify({
+          email: email,
+        }, null, {$set: {_perishable_token: token}}, (err, doc) => {
+          if (err || !doc.value) {
+            console.error(err);
+            reject(err);
+          } else {
+            doc.value._perishable_token = token;
+            resolve(doc.value);
+          }
+        });
+      });
+    });
+  }
+
+  sendPasswordResetEmail(email) {
+    if (!this.adapter) {
+      throw "Trying to send a reset password but no adapter is set";
+      //  TODO: No adapter?
+      return;
+    }
+    
+    return this.setPasswordResetToken(email).then((user) => {
+
+      const token = encodeURIComponent(user._perishable_token);
+      const username = encodeURIComponent(user.username);    
+      let link = `${this.config.requestResetPasswordURL}?token=${token}&username=${username}`
+      
+      if (!user.username) {
+        console.log('No username...');
```

Comment:
Can we avoid adding these random console.logs? Our tests already spew so much garbage...